extpsdk/README.md

```diff
@@ -0,0 +1,260 @@
+
+Query responses are responses to a `SELECT` request from Vantiq that targets a source, and can either be a `dict` or a series of `dict` objects in the form of a `list` or `array`. They only mean anything in relation to an initial Query message received from Vantiq, and thus should
+only be sent as part response to a query via the callback. 
+
+To enable this behavior, the callback is passed a _context_ parameter that provides information about the context of the query. This parameter will take the form of a `dict` object with two (2)
+items: a `source_name` item that contains a string specifying the source toward which this query is
+sent, and a `response_address` item that contains the routing information to allow the query resposnes (or error) to find their way back the the Vantiq code that issued the query.
+
+##### <a name="queryResponse" id="queryResponse"></a>Query Responses
+
+To send a query response, call
+
+```
+VantiqSourceConnection.send_query_response(ctx, code, message)
+```
+
+The `ctx` or context parameter is passed in to the `handle_query` callback.  Simply pass that back unchanged.
+
+The code is an HTTP status code. You can use the following:
+
+* `VantiqConnector.QUERY_EMPTY` to indicate that this query has an empty response.  In this case, the message parameter must be None.
+* `VantiqConnector.QUERY_PARTIAL` to indicate that this is one of a number of responses to this query. This is used if you are dynamically producing results to return.
+* `VantiqConnector.QUERY_COMPLETE` to indicate that this is the last (or only) response to this query.
+
+For the latter two (2) codes, the `message` parameter must be either a `dict` or a `list` of `dict` objects.  These will be sent to the Vantiq server in a JSON format.
+
+##### <a name="queryError" id="queryError"></a>Query Errors
+
+Query errors are sent when a Query cannot be completed successfully. To send a Query error, call
+`VantiqSourceConnection.send_query_error(ctx, message)`. 
+
+The `ctx` or query context was passed in to the `handle_query` callback.
+The `message` (`dict`) parameter must contain the following items.
+
+* `VantiqConnector.ERROR_CODE` -- the short name for the error
+* `VantiqConnector.ERROR_TEMPLATE` -- template for there error message.  Templates contain zero0-based indexed references for parameters list (_e.g._,`{0}` to refer to the first parameter, `{1}` the second, etc.).
+* `VantiqConnector.ERROR_PARAMETERS` -- a list of parameters for the template.
+
+##### <a name="publish" id="publish"></a>Published Messsages
+
+When the Vantiq server publishes an event to a source that is implemented using a connector, it is delivered via the `handle_publish` callback. As with a query message, the handler will be passed two (2) parameter:  A context and the message.
+
+The context in this case will contain the `source_name` item.  It need not contain the `response_address` item.  You can not respond to messages delivered via the `handle_publish` callback.
```

Comment:
```suggestion
The context in this case will contain the `VantiqConnector.SOURCE_NAME` item.  It will not contain the `VantiqConnector.RESPONSE_ADDRESS` item.  You can not respond to messages delivered via the `handle_publish` callback.
```